cmd/cpe/generate.go

```diff
@@ -2,80 +2,146 @@ package main
-	flag.Parse()
+	apiKey := os.Getenv(apiKeyEnvVar)
 
-	dbPath := cpe()
+	logHandler := slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo})
+	slog.SetDefault(slog.New(logHandler))
 
-	fmt.Printf("Sqlite file %s size: %.2f MB\n", dbPath, getSizeMB(dbPath))
+	if apiKey == "" {
+		log.Fatal(fmt.Sprintf("Must set %v environment variable", apiKeyEnvVar))
+	}
 
-	fmt.Println("Compressing DB...")
+	cwd, err := os.Getwd()
+	panicIf(err)
+	slog.Info(fmt.Sprintf("CWD: %v", cwd))
+
+	client := fleethttp.NewClient(fleethttp.WithTimeout(httpClientTimeout))
+	dbPath := getCPEs(client, apiKey, cwd)
+
+	slog.Info(fmt.Sprintf("Sqlite file %s size: %.2f MB\n", dbPath, getSizeMB(dbPath)))
+
+	slog.Info("Compressing DB...")
 	compressedPath, err := compress(dbPath)
-	panicif(err)
+	panicIf(err)
 
-	fmt.Printf("Final compressed file %s size: %.2f MB\n", compressedPath, getSizeMB(compressedPath))
-	fmt.Println("Done.")
+	slog.Info(fmt.Sprintf("Final compressed file %s size: %.2f MB\n", compressedPath, getSizeMB(compressedPath)))
+	slog.Info("Done.")
 }
 
 func getSizeMB(path string) float64 {
 	info, err := os.Stat(path)
-	panicif(err)
+	panicIf(err)
 	return float64(info.Size()) / 1024.0 / 1024.0
 }
 
-func cpe() string {
-	fmt.Println("Starting CPE sqlite generation...")
+func getCPEs(client common.HTTPClient, apiKey string, resultPath string) string {
+	slog.Info("Fetching CPEs from NVD...")
+
+	nvdClient, err := nvdapi.NewNVDClient(client, apiKey)
+	panicIf(err)
+
+	var cpes []cpedict.CPEItem
+	retryAttempts := 0
+
+	totalResults := 1
+	for startIndex := 0; startIndex < totalResults; {
+		cpeResponse, err := nvdapi.GetCPEs(nvdClient, nvdapi.GetCPEsParams{StartIndex: ptr.Int(startIndex)})
+		if err != nil {
+			if retryAttempts > maxRetryAttempts {
+				panicIf(err)
+			}
+			slog.Warn(fmt.Sprintf("NVD request returned error:'%v' Retrying in %v", err.Error(), waitTimeForRetry.String()))
+			retryAttempts++
+			time.Sleep(waitTimeForRetry)
+			continue
+		}
+		totalResults = cpeResponse.TotalResults
+		slog.Info(fmt.Sprintf("Got %v results", cpeResponse.ResultsPerPage))
+		startIndex += cpeResponse.ResultsPerPage
+		for _, product := range cpeResponse.Products {
+			cpes = append(cpes, convertToCPEItem(product.CPE))
+		}
+		slog.Info(fmt.Sprintf("Fetching index %v out of %v", startIndex, totalResults))
+	}
 
-	cwd, err := os.Getwd()
-	panicif(err)
-	fmt.Println("CWD:", cwd)
+	// Sanity check
+	if totalResults <= 1 || len(cpes) != totalResults {
+		log.Fatal(fmt.Sprintf("Invalid number of expected results:%v or actual results:%v", totalResults, len(cpes)))
+	}
 
-	resp, err := http.Get("https://nvd.nist.gov/feeds/xml/cpe/dictionary/official-cpe-dictionary_v2.3.xml.gz")
-	panicif(err)
-	defer resp.Body.Close()
+	slog.Info("Generating CPE sqlite DB...")
 
-	remoteEtag := getSanitizedEtag(resp)
-	fmt.Println("Got ETag:", remoteEtag)
+	dbPath := filepath.Join(resultPath, fmt.Sprint("cpe.sqlite"))
+	err = nvd.GenerateCPEDB(dbPath, cpes)
```

Comment:
TODO: Get the hash of the file and make it part of the name, so that result will be: cpe-####.sqlite